File: frontend/src/api/api-keys.ts
import { openHands } from "./open-hands-axios";

export interface ApiKey {
  id: string;
  name: string;
  prefix: string;
  created_at: string;
  last_used_at: string | null;
}

export interface CreateApiKeyResponse {
  id: string;
  name: string;
  key: string; // Full key, only returned once upon creation
  prefix: string;
  created_at: string;
}

class ApiKeysClient {
  /**
   * Get all API keys for the current user
   */
  static async getApiKeys(): Promise<ApiKey[]> {
    const { data } = await openHands.get<unknown>("/api/keys");
    // Ensure we always return an array, even if the API returns something else
    return Array.isArray(data) ? (data as ApiKey[]) : [];
  }

  /**
   * Create a new API key
   * @param name - A descriptive name for the API key
   */
  static async createApiKey(name: string): Promise<CreateApiKeyResponse> {
    const { data } = await openHands.post<CreateApiKeyResponse>("/api/keys", {
      name,
    });
    return data;
  }

  /**
   * Delete an API key
   * @param id - The ID of the API key to delete
   */
  static async deleteApiKey(id: string): Promise<void> {
    await openHands.delete(`/api/keys/${id}`);
  }
}

export default ApiKeysClient;

File: openhands/server/routes/api_keys.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from typing import List, Optional
import uuid
from datetime import datetime

from openhands.server.auth import get_current_user
from openhands.server.shared import db

# Create a router
app = APIRouter()

# Models
class ApiKey(BaseModel):
    id: str
    name: str
    prefix: str
    created_at: str
    last_used_at: Optional[str] = None

class ApiKeyCreate(BaseModel):
    name: str

class ApiKeyResponse(BaseModel):
    id: str
    name: str
    key: str
    prefix: str
    created_at: str

# Routes
@app.get("/api/keys", response_model=List[ApiKey])
async def get_api_keys(user=Depends(get_current_user)):
    """Get all API keys for the current user"""
    # In a real implementation, this would fetch from a database
    # For now, we'll return a mock empty list
    return []

@app.post("/api/keys", response_model=ApiKeyResponse)
async def create_api_key(key_data: ApiKeyCreate, user=Depends(get_current_user)):
    """Create a new API key"""
    # Generate a new API key
    key_id = str(uuid.uuid4())
    api_key = f"oh-{key_id}"
    prefix = api_key[:8]  # First 8 characters as prefix
    now = datetime.utcnow().isoformat()
    
    # In a real implementation, this would save to a database
    # For now, we'll just return the created key
    return ApiKeyResponse(
        id=key_id,
        name=key_data.name,
        key=api_key,
        prefix=prefix,
        created_at=now
    )

@app.delete("/api/keys/{key_id}")
async def delete_api_key(key_id: str, user=Depends(get_current_user)):
    """Delete an API key"""
    # In a real implementation, this would delete from a database
    # For now, we'll just return a success response
    return {"status": "success"}